src/ImageSharp/Formats/Tiff/TiffDecoderMetadataHelpers.cs

```diff
@@ -0,0 +1,114 @@
+        {
+            var coreMetadata = new ImageMetadata();
+
+            TiffFrameMetadata rootFrameMetadata = frames.First();
+            switch (rootFrameMetadata.ResolutionUnit)
+            {
+                case TiffResolutionUnit.None:
+                    coreMetadata.ResolutionUnits = PixelResolutionUnit.AspectRatio;
+                    break;
+                case TiffResolutionUnit.Inch:
+                    coreMetadata.ResolutionUnits = PixelResolutionUnit.PixelsPerInch;
+                    break;
+                case TiffResolutionUnit.Centimeter:
+                    coreMetadata.ResolutionUnits = PixelResolutionUnit.PixelsPerCentimeter;
+                    break;
+            }
+
+            if (rootFrameMetadata.HorizontalResolution != null)
+            {
+                coreMetadata.HorizontalResolution = rootFrameMetadata.HorizontalResolution.Value;
+            }
+
+            if (rootFrameMetadata.VerticalResolution != null)
+            {
+                coreMetadata.VerticalResolution = rootFrameMetadata.VerticalResolution.Value;
+            }
+
+            TiffMetadata tiffMetadata = coreMetadata.GetTiffMetadata();
+            tiffMetadata.ByteOrder = byteOrder;
+            tiffMetadata.BitsPerPixel = GetBitsPerPixel(rootFrameMetadata);
+            tiffMetadata.Compression = rootFrameMetadata.Compression;
+
+            if (!ignoreMetadata)
+            {
+                foreach (TiffFrameMetadata frame in frames)
+                {
+                    if (tiffMetadata.XmpProfile == null)
+                    {
+                        byte[] buf = frame.GetArray<byte>(ExifTag.XMP, true);
+                        if (buf != null)
+                        {
+                            tiffMetadata.XmpProfile = buf;
+                        }
+                    }
+
+                    if (coreMetadata.IptcProfile == null)
+                    {
+                        byte[] buf = frame.GetArray<byte>(ExifTag.IPTC, true);
+                        if (buf != null)
+                        {
+                            coreMetadata.IptcProfile = new IptcProfile(buf);
+                        }
+                    }
+
+                    if (coreMetadata.IccProfile == null)
+                    {
+                        byte[] buf = frame.GetArray<byte>(ExifTag.IccProfile, true);
+                        if (buf != null)
+                        {
+                            coreMetadata.IccProfile = new IccProfile(buf);
+                        }
+                    }
+                }
+            }
+
+            return coreMetadata;
+        }
+
+        private static TiffBitsPerPixel GetBitsPerPixel(TiffFrameMetadata firstFrameMetaData)
+        {
+            ushort[] bitsPerSample = firstFrameMetaData.BitsPerSample;
+            var bitsPerPixel = 0;
+            foreach (var bps in bitsPerSample)
```

Comment:
use `for` over `foreach` in arrays where possible, the codegen is better.

Reply:
Done!